text/0763-asset-importing-spec.md

```diff
@@ -0,0 +1,200 @@
++  }
+}
+```
+
+The regular expressions that broccoli-asset-rev uses to do reference rewriting are effectively un-editable at this point. Any possible change you'd make to them breaks somebody's app.
+
+Another problem with broccoli-asset-rev is that it's "push based", meaning it handles files because they exist, rather than handling files because somebody is actually consuming them. It can't detect unused assets and it can't detect missing assets at build time. It exposes the underlying implementation of how assets get delivered, making it hard to change later without breaking app code.
+
+And because broccoli-asset-rev exposes assumptions about the exact layout of the final build output, it's not compatible with Embroider. Embroider allows different Javascript packagers to optimize the app delivery in any spec-compliant way. For example, small images might get inlined as `data:` URIs, CSS `@import()` might get inlined or not, etc. All these concerns really need to be integrated into the Javascript packager, because asset handling affects app-wide hashing & fingerprinting.
+
+In contrast, a pull-based design that lets code declare what assets it needs and then not worry about how those assets will get delivered is safer and easier to change in the future.
+
+## Detailed design
+
+Inter-file references take many forms. Here is a guide to inter-file references organized by which kind of file is doing the referring ("From Type") and which type of file it's referring to ("To Type"). Only the row containing **Proposed** is new in this RFC:
+
+From Type | To Type | Example Syntax | Semantics
+--------------------|-----------------------|--------|---
+JS | JS | `import { stuff } from 'thing';`| ECMA with NodeJS resolving rules
+JS | CSS | `import 'thing/style.css';` | [Embroider Spec CSS Rule](https://github.com/emberjs/rfcs/blob/master/text/0507-embroider-v2-package-format.md#css) (1)
+JS| Other Asset (2) | `import url from 'thing/icon.png';`| **Proposed** URL Rule (3)
+CSS | CSS | `@import("thing/style.css")` | [W3C](https://drafts.csswg.org/css-cascade/#at-import) plus NodeJS resolving rules
+CSS | Other | `url("thing/icon.png");` | W3C plus NodeJS resolving rules
+HTML | JS | `<script src="./thing.js"></script>` | W3C
+HTML | CSS | `<link rel="stylesheet" href="./thing.css"/>` | W3C
+HTML | Other | `<img src="./thing.png" />` | W3C
+
+
+1. Importing a file with an explicit `.css` extension guarantees that the given CSS will be loaded into the DOM before your module executes. We do not define any exported values, this is purely for side-effect. This rule is not part of the present RFC, it was already in RFC 507.
+2. "Other Asset" means this rule applies to any JS import of a path that ends in an explicit file extension that is not `.js` or `.css`, meaning:
+    ```js
+    function isOther(theImportedPath) {
+      let extension = /\.([^.\/]+)$/.exec(theImportedPath)?.[1]?.toLowerCase();
+      return extension && !['js', 'css'].includes(extension);
```

Comment:
any reason to exclude ts from here?

Reply:
I suppose this same caveat/note [here](https://github.com/emberjs/rfcs/pull/763/files#diff-38fadf17712b2606dfa948983f8d8d89d28c66e71b346d7c7b5bccc5808bf446R93) applies to ts

Reply:
I would say the "Extensibility" section applies. Compiles-to-javascript languages aren't covered by this proposal because we're only standardizing what happens in the Javascript.

Also, TypeScript in particular forbids an explicit `.ts` extension in its import statements, or at least it used to. They rely on the ambiguity to keep the import valid after everything has converted from ts to js.